src/esw/imu_driver.py

```diff
@@ -63,30 +137,46 @@ def main():
             rospy.logerr("incomplete msg")
             continue
 
+        # rotate the quaternion and mag vector by 90 degrees about the Z axis to convert it to ENU frame
+        offset_quat = quaternion_about_axis(np.pi / 2, [0, 0, 1])
+        enu_orientation = quaternion_multiply(offset_quat, orientation_data)
+        R = rotation_matrix(np.pi / 2, [0, 0, 1])
+        enu_mag_vec = R @ np.array(mag_data)
+
         # fill in all sensor messages, setting timestamps of each message to right now,
         # and setting the reference frame of all messages to IMU frame
-        imu_msg = ImuAndMag(
-            header=Header(stamp=rospy.Time.now(), frame_id=imu_frame),
-            imu=Imu(
-                header=Header(stamp=rospy.Time.now(), frame_id=imu_frame),
-                orientation=Quaternion(*orientation_data),
+        header = Header(stamp=rospy.Time.now(), frame_id=imu_frame)
+        mag_msg = (
+            MagneticField(
+                header=header,
+                magnetic_field=Vector3(*enu_mag_vec),
+            ),
+        )
+        imu_msg = (
+            Imu(
+                header=header,
+                orientation=Quaternion(*enu_orientation),
                 linear_acceleration=Vector3(*accel_data),
                 angular_velocity=Vector3(*gyro_data),
             ),
-            mag=MagneticField(
-                header=Header(stamp=rospy.Time.now(), frame_id=imu_frame),
-                magnetic_field=Vector3(*mag_data),
-            ),
+        )
+        imu_msg = inject_covariances(imu_msg, orientation_covariance, gyro_covariance, accel_covariance)
```

Comment:
The reason I don't do that is to minimize code duplication between this and the sim imu driver, which takes in an existing Imu message and only needs to add the covariance matrices to it

Reply:
hm I see. You could also keep the function for use in the sim imu driver but just construct the IMU message here with all the fields so you're not making a copy every time. like: 

```
imu_msg = Imu(
                header=header,
                orientation=Quaternion(*enu_orientation),
                linear_acceleration=Vector3(*accel_data),
                angular_velocity=Vector3(*gyro_data),
                orientation_covariance = ...
                angular_velocity_coraiance = ...
                ...
            )
```

This way the function would I guess only be used once but it would only be used when you actually are augmenting an IMU message that you're ingesting as opposed to just always even when you're making the IMU message yourself.

Reply:
is your concern here performance or style? because I thought we agreed from one of the other comments that it's ok to sacrifice the performance of a copy in order to have a pure function, and the performance probably won't matter anyway.

Reply:
Yea I mean I guess the performance doesn't really matter here but I feel like its a bit awkward to create an object and then pass it to a function and replace it immediately when you can just fill in all the fields with the constructor. I see the purpose of the function but think it should be moved to sim_imu_driver and within imu_driver think you should just construct the object with all of the fields since you already have all of them

Reply:
yeah I guess this is such a small piece of code that it's probably ok to just duplicate instead of trying to abstract it. will do